detectors/aws/ec2/v2/ec2_test.go

```diff
@@ -0,0 +1,144 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ec2
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"testing"
+	"time"
+
+	"github.com/aws/smithy-go/middleware"
+
+	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
+)
+
+type mockClient struct {
+	mock.Mock
+}
+
+func (m *mockClient) GetInstanceIdentityDocument(ctx context.Context, params *imds.GetInstanceIdentityDocumentInput, optFns ...func(*imds.Options)) (*imds.GetInstanceIdentityDocumentOutput, error) {
+	args := m.Called(ctx, params, optFns)
+	return args.Get(0).(*imds.GetInstanceIdentityDocumentOutput), args.Error(1)
+}
+
+func (m *mockClient) GetMetadata(ctx context.Context, params *imds.GetMetadataInput, optFns ...func(*imds.Options)) (*imds.GetMetadataOutput, error) {
+	args := m.Called(ctx, params, optFns)
+	return args.Get(0).(*imds.GetMetadataOutput), args.Error(1)
+}
+
+func TestAWS_ResourceDetection(t *testing.T) {
+	doc := validIdentityDocument()
+
+	testCases := []struct {
+		name           string
+		metadataOutput *imds.GetMetadataOutput
+		metadataErr    error
+		docOutput      *imds.GetInstanceIdentityDocumentOutput
+		docErr         error
+		expectedAttrs  []attribute.KeyValue
+	}{
+		{
+			name:           "AllFields",
+			docOutput:      doc,
+			metadataOutput: mockMetadataOutput("ip-12-34-56-78.us-west-2.compute.internal"),
+			expectedAttrs: []attribute.KeyValue{
+				semconv.CloudProviderAWS,
+				semconv.CloudPlatformAWSEC2,
+				semconv.CloudRegion("us-west-2"),
+				semconv.CloudAvailabilityZone("us-west-2b"),
+				semconv.CloudAccountID("123456789012"),
+				semconv.HostID("i-1234567890abcdef0"),
+				semconv.HostImageID("ami-5fb8c835"),
+				semconv.HostType("t2.micro"),
+				semconv.HostName("ip-12-34-56-78.us-west-2.compute.internal"),
+			},
+		},
+		{
+			name:           "NoHostname",
+			docOutput:      doc,
+			metadataOutput: mockMetadataOutput(""),
+			metadataErr:    errors.New("mock error"),
+			expectedAttrs: []attribute.KeyValue{
+				semconv.CloudProviderAWS,
+				semconv.CloudPlatformAWSEC2,
+				semconv.CloudRegion("us-west-2"),
+				semconv.CloudAvailabilityZone("us-west-2b"),
+				semconv.CloudAccountID("123456789012"),
+				semconv.HostID("i-1234567890abcdef0"),
+				semconv.HostImageID("ami-5fb8c835"),
+				semconv.HostType("t2.micro"),
+			},
+		},
+		{
+			name:           "NonEC2Host",
+			docErr:         errors.New("error getting InstanceIdentityDocument"),
+			docOutput:      &imds.GetInstanceIdentityDocumentOutput{},
+			metadataOutput: mockMetadataOutput(""),
+			expectedAttrs:  nil, // Empty resource
+		},
+	}
```

Comment:
Could you add tests cases for 404, and another HTTP status code error (probably 500)? So those cases in recordError are covered?